src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs

```diff
@@ -0,0 +1,94 @@
+        /// <summary>The number of items in <see cref="_array"/>, and thus also the next position in the array to be filled.</summary>
+        private int _count;
+
+        /// <summary>Initializes the builder.</summary>
+        /// <param name="capacity">The initial capacity of the builder.</param>
+        public SegmentStringBuilder(int capacity)
+        {
+            Debug.Assert(capacity > 0);
+            _array = ArrayPool<ReadOnlyMemory<char>>.Shared.Rent(capacity);
+            _count = 0;
+        }
+
+        /// <summary>Gets the number of segments added to the builder.</summary>
+        public int Count => _count;
+
+        /// <summary>Adds a segment to the builder.</summary>
+        /// <param name="segment">The segment.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Add(ReadOnlyMemory<char> segment)
+        {
+            ReadOnlyMemory<char>[] array = _array;
+            int pos = _count;
+            if ((uint)pos < (uint)array.Length)
+            {
+                array[pos] = segment;
+                _count = pos + 1;
+            }
+            else
+            {
+                GrowAndAdd(segment);
+            }
+        }
+
+        /// <summary>Grows the builder to accomodate another segment.</summary>
+        /// <param name="segment"></param>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void GrowAndAdd(ReadOnlyMemory<char> segment)
+        {
+            ReadOnlyMemory<char>[] array = _array;
+            Debug.Assert(array.Length == _count);
+
+            ReadOnlyMemory<char>[] newArray = _array = ArrayPool<ReadOnlyMemory<char>>.Shared.Rent(array.Length * 2);
+            Array.Copy(array, newArray, _count);
+            ArrayPool<ReadOnlyMemory<char>>.Shared.Return(array, clearArray: true);
+            newArray[_count++] = segment;
+        }
+
+        /// <summary>Gets a span of all segments in the builder.</summary>
+        /// <returns></returns>
+        public Span<ReadOnlyMemory<char>> AsSpan() => new Span<ReadOnlyMemory<char>>(_array, 0, _count);
+
+        /// <summary>Creates a string from all the segments in the builder and then disposes of the builder.</summary>
+        public override string ToString()
+        {
+            int length = 0;
+            foreach (ReadOnlyMemory<char> segment in AsSpan())
+            {
+                length += segment.Length;
+            }
+
+            string result = string.Create(length, this, (dest, builder) =>
+            {
+                foreach (ReadOnlyMemory<char> segment in builder.AsSpan())
+                {
+                    segment.Span.CopyTo(dest);
+                    dest = dest.Slice(segment.Length);
+                }
+            });
+
+            ReadOnlyMemory<char>[] array = _array;
+            AsSpan().Clear(); // clear just what's been filled
+            this = default;
+            ArrayPool<ReadOnlyMemory<char>>.Shared.Return(array);
```

Comment:
`_array = null;` ?

Reply:
The `this = default;` is doing that.

Reply:
Oh, I didn't acutally know you could do that.